app/src/main/java/com/github/epfl/meili/messages/ChatLogActivity.kt

```diff
@@ -0,0 +1,119 @@
+import android.widget.EditText
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import com.github.epfl.meili.models.ChatMessage
+import com.github.epfl.meili.models.User
+import com.google.firebase.auth.FirebaseAuth
+import com.google.firebase.database.ChildEventListener
+import com.google.firebase.database.DataSnapshot
+import com.google.firebase.database.DatabaseError
+import com.google.firebase.database.FirebaseDatabase
+import com.xwray.groupie.GroupAdapter
+import com.xwray.groupie.GroupieViewHolder
+import com.xwray.groupie.Item
+
+class ChatLogActivity : AppCompatActivity() {
+
+
+    companion object {
+        val TAG = "ChatLog"
+    }
+
+    val adapter = GroupAdapter<GroupieViewHolder>()
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_chat_log)
+
+        findViewById<RecyclerView>(R.id.recycleview_chat_log).adapter = adapter
+
+        val user = intent.getParcelableExtra<User>(NewMessageActivity.USER_KEY)
+        supportActionBar?.title = user?.username
+
+        listenForMessages()
+
+        findViewById<Button>(R.id.button_chat_log).setOnClickListener {
+            performSendMessage()
+        }
+
+
+    }
+
+    private fun performSendMessage() {
+        val text = findViewById<EditText>(R.id.edit_text_chat_log).text.toString()
+
+        val fromId = FirebaseAuth.getInstance().uid
+        val user = intent.getParcelableExtra<User>(NewMessageActivity.USER_KEY)
+        val toId = user?.uid
+
+        if (fromId == null) return
+        if (toId == null) return
+
+        val reference = FirebaseDatabase.getInstance().getReference("/messages").push()
+
+        val chatMessage =
+            ChatMessage(reference.key!!, text, fromId, toId, System.currentTimeMillis() / 1000)
+        reference.setValue(chatMessage)
+            .addOnSuccessListener {
+                Log.d(TAG, "Saved our chat message: ${reference.key}")
+            }
+    }
+
+    private fun listenForMessages() {
+        val user = intent.getParcelableExtra<User>(NewMessageActivity.USER_KEY)
+        val toId = user?.uid
+
+        val ref = FirebaseDatabase.getInstance().getReference("/messages")
+        ref.addChildEventListener(object : ChildEventListener {
+            override fun onChildAdded(snapshot: DataSnapshot, previousChildName: String?) {
+                val chatMessage = snapshot.getValue(ChatMessage::class.java)
+
+                if (chatMessage != null) {
+                    if (chatMessage.fromId == FirebaseAuth.getInstance().uid) {
+                        adapter.add(ChatItem(chatMessage.text, true))
+                    } else if (chatMessage.toId == toId) {
+                        adapter.add(ChatItem(chatMessage.text, false))
+                    }
+                }
+            }
+
+            override fun onChildChanged(snapshot: DataSnapshot, previousChildName: String?) {
```

Comment:
Unfortunately yes since the ChildEventListener is an abstract class I have to override all the methods. I tried finding workarounds but this seems to be the go-to solution on stack-overflow.